### ISSUE TITLE
BLIT image function fails 

### ISSUE THREAD

**Issue by Arcanister** [wesnoth/wesnoth#2058]
I have been experimenting with add-on publishing facility and get this message:

error config: Failed to apply a modification to an image:
Image: attacks/thorns.png.
Modifications: ~BLIT(icons/icon-addon-publish.png).
Error: ~BLIT(): offset and width '64' larger than destination image's width '60' no blitting performed.

As far, as I understand, this happens because of [addon_list.cpp:151](https://github.com/wesnoth/wesnoth/blob/master/src/gui/widgets/addon_list.cpp#L151).
Add-on icon is supposed to be displayed with green + on top when it is publishable, my icon [thorns.png](https://github.com/wesnoth/wesnoth/blob/master/data/core/images/attacks/thorns.png) was 60x60, while  [icon-addon-publish.png](https://github.com/wesnoth/wesnoth/blob/master/images/icons/icon-addon-publish.png) is 64x64.

A solution would be to teach BLIT to recognize negative coordinates as coming from low-right corner and make publish icon as small as necessary to fit the cross, then use something like ~BLIT(small-icon-publish.png,-16,-16).
Another solution would be to add some other image path function in front of ~BLIT(), which expands add-on icon to desired size, such as 64x64 and apply BLIT after that.

**Comment by doofus-01** [comment]
Hi @Arcanister ,
I don't see this as an issue, really.
> Another solution would be to add some other image path function in front of ~BLIT(), which expands add-on icon to desired size, such as 64x64 and apply BLIT after that.

You could just blit both images onto misc/blank-hex.png, and scale as needed.

It is intuitive that you can't fit a larger image into a smaller image.  Scaling the larger image first, or not using the smaller image as a base, would fix it.  I'd rather have the error than have the function switch to some other behavior when I don't use it correctly.

**Comment by CelticMinstrel** [comment]
I don't think it actually applies to this case, but allowing negative coordinates in BLIT _does_ sound like a nice idea.

**Comment by Arcanister** [comment]
@doofus-01 The issue is not in my add-on (which is really a simple test case),  the problem is in Wesnoth main code which tries to BLIT() a larger image (+ icon) onto smaller image (user-choosen icon).

**Comment by doofus-01** [comment]
@Arcanister - OK, then I'd still say BLIT() is behaving just fine, the issue is with the add-on manager GUI.  Without testing, I won't speculate what exactly should be done, but changing the way BLIT() behaves surely isn't the answer.

**Comment by Arcanister** [comment]
Sure, it's not an answer **by itself**. I have changed initial message to clarify it, but I did mean all this from beginning.

**Comment by CelticMinstrel** [comment]
I dunno, it does seem to me that automatically clipping to the destination image's bounds would have made sense for BLIT though.

**Comment by doofus-01** [comment]
@Arcanister - I'd still say there is nothing wrong with BLIT().  Your second solution seems reasonable, but I don't know why it should be 64x64px.  72px might be better, though even that might be problematic for people who want to use bigger images.  Making BLIT( ) auto-detect the largest image might be simple, I have no idea, but it seems unnecessary;  at the end of the day, you chose an image that didn't work, you got an error, and now you can easily fix it so that it does work.  So, I'd still not be convinced anything is wrong here.

@CelticMinstrel - It might be nice in some circumstances, but I'd still rather have the error, since it meant I wasn't paying attention to the coordinates on the base image or the size of the applied image.  For an image that is built up by macros, this could be an issue, silently leading to messed up images I might not see for a while.
